Initial maven support

jroper opened the pull request:
This is initial support for maven. The main thing here is that the Lagom development environment is working. Cassandra gets started, the service locator gets started, every project gets started and configured correctly with dev mode dependencies injected, and hot reloads work.

I've tested this with the helloworld Lagom seed project, which I've converted to use maven here:

https://github.com/jroper/lagom-helloworld-maven

The general approach to implementation has been:

* Extract and abstract common utilities into lagom-build-tool-support
* Create lagom-maven-plugin
* I've manually created the maven META-INF/maven/plugin.xml file, when using maven to develop a maven plugin, this file is usually generated from annotations. I investigated whether it was feasible to reuse the maven generation tool, the amount of code in the ant plugin to bootstrap maven itself in order to run this plugin doesn't make it look like it's worth it.
* Most of the interactions with maven happen in the MavenFacade class, this includes executing lifecycle phases such as compile, and resolving dependencies.
* I've generally stuck to using quite low level maven APIs, for example, it's really trivial to execute a lifecycle phase on a particular project and all its upstream dependencies if using the maven execution API, however using this means each time its run, pom files have to be reparsed, dependencies reresolved, each time you run it, which is slow. So instead, to run a compile for example, I get all the projects I need to run it on, then look up all the tasks that make up the compile lifecycle, then run each one of them.

There's a lot left to do, here's an incomplete list:

- [x] Create some sort of test environment for testing the maven plugin, and write tests.
- [x] Source map and compile error handling (not sure if possible).
- [x] Emebbeded Play project support (needed for chirper).
- [x] External project support.
- [x] Configurable dev settings.
- [x] Documentation.
- [x] Generate the parts of plugin.xml that are dependent on the build/version etc.
- [x] Create maven archetypes for helloworld and chirper.

dotta commented on dev/build-tool-support/src/main/scala/com/lightbend/lagom/dev/Reloader.scala:
```diff
@@ -159,26 +160,46 @@ class Reloader(
-    import scala.collection.JavaConverters._
-    listeners.iterator().asScala.foreach(listener => listener())
+    val now = Instant.now()
+    fileLastChanged.set(now)
```
There is a possible race condition here. (it's too bad there isn't an atomic `setAndGet` method)

jroper replied:
I don't think so. Every time there is a change, the file last changed atomic reference is updated and a timer task is then set. That timer task uses a compareAndSet to set it to null if the reference hasn't changed since it was set, and if the compare condition holds true, ie, there have been no changes during the quiet period), then a reload is done.

dotta replied:
> if the compare condition holds true, ie, there have been no changes during the quiet period

This statement isn't true. Say you have two threads t1,t2. t1 gets to the line that updates `fileLastChanged`. Then, t2 arrives and updates `fileLastChanged` as well. Now, when the quiet period set by t1 finishes, the expression `fileLastChanged.compareAndSet(start, null)` will return `false`, indicating that there was no change during the quiet period, which is wrong.

But now that I think more about it, the race condition can be probably qualified of "benign", as `fileLastChanged.compareAndSet(start, null)` will eventually yield true (because of the update made by t2 to `fileLastChanged`).

Therefore, I think we are good.

jroper replied:
> the expression fileLastChanged.compareAndSet(start, null) will return false, indicating that there was no change during the quiet period, which is wrong.

That's exactly the point, and it's exactly what's supposed to happen, returning false doesn't mean there were no changes, it means there *were* changes, which tells it that it wasn't quiet during the quiet period, which means there was no quiet period, which means it shouldn't execute.  What I'm trying to prevent here is a situation where the IDE saves 20 files in a short time, that triggers 20 file changed notifications, which causes the app to recompile 20 times.  Because sbt has good incremental compilation support, we never even noticed that this was happening before, but with maven, it's really slow and causes a big problem.  So what we want to do is ensure that only the *last* file changed notification in batch of notifications causes a recompile, and we do that using this quite period timeout mechanism.

dotta replied:
👍

dotta commented on dev/build-tool-support/src/main/scala/com/lightbend/lagom/dev/Reloader.scala:
```diff
@@ -193,6 +214,7 @@ class Reloader(
    * - null - If nothing changed.
    */
   def reload: AnyRef = {
+    // fixme: synchronize on something that is guaranteed to be global in mavens multi classloader environment
```
Are you planning to fix this before merging? Just pointing it out in case you have forgot about your own note.

jroper replied:
Maybe - there's a lot of todos/fixmes.

dotta replied:
I'd say we should try to resolve all fixme/todo before merging. Either by fixing them, or creating tickets. (I think it'd be valuable to have a policy that is applicable to all contributions.)

dotta commented on dev/maven-plugin/src/main/resources/META-INF/maven/plugin.xml:
```diff
@@ -0,0 +1,384 @@
+    <description>Provides maven support for Lagom Framework</description>
+    <groupId>com.lightbend.lagom</groupId>
+    <artifactId>lagom-maven-plugin</artifactId>
+    <version>1.1.0-SNAPSHOT</version>
```
I guess this should be 1.0.1-SNAPSHOT - unless you plan on generating this file before merging.

dotta replied:
Disregard this comment, I just noticed https://github.com/lagom/lagom/pull/137/files#diff-daa348ddc8d0bbd06f657bb9e2de314aR1

dotta commented on dev/maven-plugin/src/main/resources/META-INF/maven/plugin.xml:
```diff
@@ -0,0 +1,384 @@
+            <parameters>
+                <parameter>
+                    <name>cassandraMaxBootWaitingSeconds</name>
+                    <type>int</type>
```
It's a `Long` on the sbt plug-in.

dotta commented on dev/maven-plugin/src/main/scala/com/lightbend/lagom/maven/ConfigureMojo.scala:
```diff
@@ -0,0 +1,23 @@
+class ConfigureMojo @Inject() (session: MavenSession) extends AbstractMojo {
+
+  @BeanProperty
+  var lagomService: Boolean = _
```
I'm wondering if this may be accessed from different threads. If yes, it needs to be `volatile`

jroper replied:
It can't be, mojos aren't passed around. And, since it's not a singleton, it will be instantiated for each time it's used, and it's use is only ever in a single threaded manner.

dotta replied:
Awesome. Thanks for the background info!

dotta commented on dev/maven-plugin/src/main/scala/com/lightbend/lagom/maven/ScalaClassLoaderManager.scala:
```diff
@@ -0,0 +1,63 @@
+    }.sorted.mkString(",")
+  }
+
+  private var cache = Map.empty[String, ClassLoader]
```
I wonder if this could lead to classloader leakages

jroper replied:
Yes and no. There's no maven console, you don't start maven, run it in dev mode, stop it, run tests, run it in dev mode again, etc, all in the one process.  Unlike sbt, in maven you only ever do one thing, then the process terminates.  So classloader leaks are not nearly as much of a problem.

dotta replied:
👍

dotta commented on dev/maven-plugin/src/main/scala/com/lightbend/lagom/maven/ServerMojos.scala:
```diff
@@ -0,0 +1,91 @@
+      // Configure logging to quieten the Cassandra driver
+      mavenLoggerManager.getLoggerForComponent("com.datastax").setThreshold(Logger.LEVEL_DISABLED)
+
+      val cp = facade.resolveArtifact(new DefaultArtifact("com.lightbend.lagom", "lagom-cassandra-server_2.11",
```
Might be good to extract the appended Scala version somewhere so that it will be easy to support both 2.11 and 2.12 in the near future.

jroper replied:
Actually, lagom-cassandra-server doesn't even depend on Scala (or at least it shouldn't), since it's just some configuration files and a dependency on cassandra itself. What we should probably do is turn off cross versioning for it.

dotta replied:
I see your point, but it does (indirectly) require a specific Scala version. I've trouble seeing what there is to gain by removing the suffix.

dotta commented on dev/maven-plugin/src/main/scala/com/lightbend/lagom/maven/ServiceManager.scala:
```diff
@@ -0,0 +1,173 @@
+  }
+
+  def stopService(project: MavenProject) = {
+    runningServices.get(project) match {
```
`runningServices` needs to be read either by holding the instance lock, or it should be declared `volatile` to avoid visibility issues.

jroper replied:
Good catch.

dotta commented:
Looks good!

The one thing I'm wondering about is the Maven threading model, and if the POJO classes may be executed from different threads. If so, I believe we will need (at a minimum) to have `volatile` added to the beans' properties.

dotta commented on build.sbt:
```diff
@@ -10,11 +10,12 @@ import de.heikoseeberger.sbtheader.HeaderPattern
 val PlayVersion = "2.5.0"
 val AkkaVersion = "2.4.4"
-val AkkaPersistenceCassandraVersion = "0.13"
+val AkkaPersistenceCassandraVersion = "0.17"
```
This PR https://github.com/lagom/lagom/pull/134 (which was merged already) takes care of updating the akka-persistence-cassandra version

jroper commented on dev/maven-plugin/src/main/maven/plugin.xml:
```diff
@@ -0,0 +1,310 @@
+<!-- This is not the final file that gets used, this is processed by the SbtMavenPlugin, in the sbt project directory. -->
```
@dotta Checkout the new plugin.xml, not only are properties that can be determined from the build generated, but I've created a parameterManagement section that saves on the duplication of parameter definitions for each mojo.

jroper commented on project/SbtMavenPlugin.scala:
```diff
@@ -0,0 +1,228 @@
+package lagom
```
@dotta And this is the sbt plugin that processes the new plugins.xml.

dotta commented on dev/build-tool-support/src/main/scala/com/lightbend/lagom/dev/ConsoleHelper.scala:
```diff
@@ -0,0 +1,84 @@
+      val closing = Future.traverse(services)(serv => Future {
+        serv.close()
+      })
+      closing.onComplete(_ => println("All services are stopped"))
```
change `println` w/ `log` statement (?)

jroper replied:
Sure.

dotta commented on dev/build-tool-support/src/main/scala/com/lightbend/lagom/dev/ConsoleHelper.scala:
```diff
@@ -0,0 +1,84 @@
+class Colors(logNoFormat: String) {
+  import scala.Console._
+
+  lazy val isANSISupported = {
```
Any reason why this is a `lazy val` and not just a `val`?

jroper replied:
I don't know...

dotta commented on dev/build-tool-support/src/main/scala/com/lightbend/lagom/dev/LagomProcess.scala:
```diff
@@ -0,0 +1,30 @@
+
+    val classpathString = classpath.map(_.getAbsolutePath).mkString(File.pathSeparator)
+
+    val command = javaBin :: jvmArgs ::: "-classpath" :: classpathString :: main :: args
```
How come there are no spaces needed between the different arguments (e.g., "-classpath" and the `classpathString`).

dotta commented on dev/maven-plugin/src/main/scala/com/lightbend/lagom/maven/LagomKeys.scala:
```diff
@@ -0,0 +1,43 @@
+  val LagomServiceUrl = new LagomKey[String]("lagomServiceUrl")
+}
+
+final class LagomKey[A](name: String) {
```
Seems like this could be a value object (hence, a case class).

jroper replied:
I don't think they'll be compared, patterned matched, etc, I don't see the value in the byte code overhead.  I probably should make it private.

dotta commented on dev/maven-plugin/src/main/scala/com/lightbend/lagom/maven/MavenFacade.scala:
```diff
@@ -0,0 +1,276 @@
+class MavenFacade @Inject() (repoSystem: RepositorySystem, session: MavenSession,
+    buildPluginManager: BuildPluginManager, lifecycleExecutionPlanCalculator: LifecycleExecutionPlanCalculator) {
+
+  val cachedProjectIndex = new ProjectIndex(session.getProjects)
```
This doesn't seem to be used anywhere (??)

jroper replied:
You're right - it's a remnant from when I was trying to do some things a different way.

dotta commented on dev/maven-plugin/src/main/scala/com/lightbend/lagom/maven/MavenFacade.scala:
```diff
@@ -0,0 +1,276 @@
+      if (!(
+        StringUtils.isEmpty(dep.getGroupId) ||
+        StringUtils.isEmpty(dep.getArtifactId) ||
+        StringUtils.isEmpty(dep.getVersion)
```
Maybe worth extracting a method here. Maybe the method's name could be something like `isValidArtifactName`?

jroper replied:
I'd rather leave as it is - this code was copied/converted to Scala from the class mentioned in the comments above, by leaving it as it is it makes it easier to correlate back to the original source it was copied from in case we have any problems and want to see how Maven does it.

dotta commented on dev/maven-plugin/src/main/scala/com/lightbend/lagom/maven/MavenFacade.scala:
```diff
@@ -0,0 +1,276 @@
+          sys.error(s"${LagomKeys.LagomService} not set on project ${project.getArtifactId} after running configure!")
+        }
+      } else {
+        // Lagom plugin not configured, return false
```
Do we want to add a log.debug statement here for making it easier to understand why their defined project isn't picked up as a service?

jroper replied:
Will do.

dotta commented on dev/maven-plugin/src/main/scala/com/lightbend/lagom/maven/PortRangeBean.scala:
```diff
@@ -0,0 +1,8 @@
+
+import scala.beans.BeanProperty
+
+class PortRangeBean {
```
This is going to be a singleton, or am I wrong? I'm wondering because I'd expect the min/max may be write/read from different threads (and, if so, we will need to establish some sort of memory barrier to avoid race conditions or visibility issues).

dotta replied:
Aha, maybe I see. The bean property setters are used to change the default when an instance of `PortRangeBean` is created. If it's guaranteed that `min/max` are written and read by the same thread, then we are good and you can disregard my above comment. Though, I wonder what would be the interest of having the same thread writing and reading values from this instance, which is why I'm being a bit suspicious here.

jroper replied:
It's a Java bean that gets instantiated and populated by maven when configuring mojos.  It will only be accessed from one thread.

dotta commented on dev/maven-plugin/src/main/scala/com/lightbend/lagom/maven/ServiceMojos.scala:
```diff
@@ -0,0 +1,199 @@
+
+    services.foreach { project =>
+      facade.executeMavenPluginGoal(project, "start")
+    }
```
What happens if starting a service fails?

jroper replied:
An exception is thrown and maven stops.

dotta commented on dev/sbt-plugin/src/main/scala/com/lightbend/lagom/sbt/LagomPlugin.scala:
```diff
@@ -285,18 +279,43 @@ object LagomPlugin extends AutoPlugin {
   ))
 
+  private val projectPortMap = AttributeKey[Map[ProjectName, Port]]("lagomProjectPortMap")
+  private val defaultPortRange = PortRange(20000, 30000)
```
Shouldn't this be the same as in https://github.com/lagom/lagom/pull/137/files#diff-d142c6e1820bd14ed4be0ccdbb3444e7R6 (I guess you can just take the opportunity to make the change and fix #101

dotta commented on testkit/src/main/scala/com/lightbend/lagom/javadsl/testkit/ServiceTest.scala:
```diff
@@ -199,9 +200,10 @@ object ServiceTest {
+        FileUtils.deleteRecursiveOnExit(cassandraDirectory)
         val t0 = System.nanoTime()
-        CassandraLauncher.start(cassandraDirectory, CassandraLauncher.DefaultTestConfigResource, clean = true, port = 0)
+        CassandraLauncher.start(cassandraDirectory, CassandraLauncher.DefaultTestConfigResource, clean = false, port = 0)
```
Interesting, why did you set `clean` to be false?

jroper replied:
Because if the directory doesn't exist, it throws an exception. The launcher attempts to create the directory after attempting to delete the directory, the directory deletion fails if the directory didn't already exists.  I'm not sure how or if this has ever worked - I think in our tests we create the directory first, I don't think we should expect end users to do that.

The other problem with the old code was that it used the target directory is the working directory - but in sbt, and in general, you shouldn't depend on the current working directory in tests, in some cases the target directory would be the parent projects target directory.  So, what I did was to use the Java 7 Files.createTemporaryDirectory to obtain a directory.  Since this is guaranteed to be a new empty directory, there's no need to clean it first.

dotta commented on dev/build-tool-support/src/main/scala/com/lightbend/lagom/dev/Reloader.scala:
```diff
@@ -138,6 +139,55 @@ object Reloader {
+
+    val _buildLink = new BuildLink {
+      private val initialized = new java.util.concurrent.atomic.AtomicBoolean(false)
+      override def runTask(task: String): AnyRef = throw new UnsupportedOperationException("Run task not support in Lagom")
```
supported

jroper commented:
This is now finished, I've updated the PR from all the review comments, I've implemented/fixed all the todos/fixmes, and I've fixed the build.  Please review and if ok, merge.

dotta commented on .travis.yml:
```diff
@@ -4,9 +4,11 @@ jdk:
-    - SCRIPT="sbt -jvm-opts .travis-jvmopts publishLocal scripted"
-    - SCRIPT="sbt -jvm-opts .travis-jvmopts unidoc && cd docs && sbt -jvm-opts .travis-jvmopts markdownValidateDocs test markdownEvaluateSbtFiles"
+    - SCRIPT="sbt --warn -jvm-opts .travis-jvmopts test"
+    # Run scripted in a separate process without --warn so we can see its progress
```
I don't understand the comment. --warn is still present...

jroper replied:
Look carefully, sbt is invoked twice (separated by &&, this gets evaled by bash so && is a delimiter for multiple commands, that will only execute the second command if the first was successful).  The first one is invoked with --warn, sbt goes very quiet when you do this, in fact it can be a problem because travis will time the build out if it doesn't log anything.  The second invocation of sbt, the one that executes scripted, doesn't use --warn.

dotta replied:
Oh, jeeez, I missed the && :)

dotta commented on build.sbt:
```diff
@@ -14,8 +14,26 @@ val AkkaPersistenceCassandraVersion = "0.17"
+// down the tree brings in netty-http-codec 4.0.36 which brings in netty-handler 4.0.36, and we also depend on
+// netty-reactive-streams, which brings in netty-handler 4.0.33, since, that's much higher in the tree than the other
+// path to it, we 4.0.33, and netty-http-codec 4.0.36 is incompatible with netty-handler 4.0.33, and so, well,
+// thankyou maven.
```
Wow - didn't know this is how Maven resolved ambiguos artefact's version. The comment is extremely useful, but I definitely see this as a can of worm, and we should be extremely cautious when upgrading versions of any dependency in the future.

dotta commented on dev/archetypes/maven-java/src/main/resources/archetype-resources/__service1Name__-impl/src/bundle/bundle.conf:
```diff
@@ -0,0 +1,28 @@
+compatibilityVersion = "1"
+system = "${service1Name}"
+systemVersion = "1"
+nrOfCpus = 0.1
```
0.1?

jroper replied:
Yeah I know.  This is the default that sbt-conductr generates, I assume the ConductR team knew what they were doing when they wrote that, so I kept it as is.

dotta commented on dev/maven-plugin/src/main/scala/com/lightbend/lagom/maven/RenameConductRBundleMojo.scala:
```diff
@@ -0,0 +1,72 @@
+        s"that either that artifact is located at ${sourceConductRBundle.getAbsolutePath}, or you have configured " +
+        s"sourceConductRBundle to point to it. If you are using the maven assembly plugin to produce the artifact, " +
+        s"ensure it is declared in your POM before the lagom plugin, as Maven executions for the same phase are " +
+        s"executed in the order that they appear in the POM.")
```
There are a few unneeded usages of string interopolation. Pointing it out in case you want to remove them.

dotta commented:
Good stuff James! The few additional comments I had are for minor things, feel free to decide what to do with them and merge at your will!
